File: sdks/python/assinavelox/_http.py
"""Transporte HTTP do SDK (urllib da biblioteca padrão). Uso interno."""

from __future__ import annotations

import json
import platform
import re
import socket
import urllib.error
import urllib.parse
import urllib.request
import uuid
from dataclasses import dataclass
from typing import Any, Iterable, Mapping

from ._version import API_MAJOR, SDK_VERSION
from .errors import ApiError, InvalidRequestError, NetworkError, RequestTimeoutError
from .files import FileUpload

IDEMPOTENCY_KEY = re.compile(r"[\x21-\x7e]{1,255}")
_UNSAFE_HEADER = re.compile(r"[\r\n\x00]")


@dataclass(frozen=True)
class RawResponse:
    status: int
    headers: dict[str, str]
    body: bytes

    def json(self) -> Any:
        return json.loads(self.body.decode("utf-8")) if self.body else None


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    """Redirecionamento não é seguido: o token nunca vai para outro endereço."""

    def redirect_request(self, req, fp, code, msg, headers, newurl):  # type: ignore[no-untyped-def]
        return None


def _scalar(value: Any) -> str:
    if isinstance(value, bool):
        return "true" if value else "false"
    return str(value)


def _quote_param(value: str) -> str:
    return value.replace('"', "%22").replace("\r", "%0D").replace("\n", "%0A")


class HttpClient:
    def __init__(
        self,
        base_url: str,
        token: str,
        timeout: float = 30.0,
        headers: Mapping[str, str] | None = None,
    ) -> None:
        if not isinstance(base_url, str) or not re.match(r"^https?://[^/\s]+", base_url):
            raise InvalidRequestError("base_url precisa ser http(s)://…/api/v1 da sua instalação.")
        if not isinstance(token, str) or not token or re.search(r"\s", token):
            raise InvalidRequestError("token vazio ou com espaço: use o texto exibido na criação da chave.")
        if timeout <= 0:
            raise InvalidRequestError("timeout precisa ser maior que zero.")
        self.base_url = base_url.rstrip("/")
        self.__token = token
        self.timeout = float(timeout)
        self.default_headers = self._check_headers(headers or {})
        self.user_agent = f"assinavelox-python/{SDK_VERSION} (api-v{API_MAJOR}; python/{platform.python_version()})"
        self._opener = urllib.request.build_opener(_NoRedirect)

    def __repr__(self) -> str:
        return f"HttpClient(base_url={self.base_url!r})"

    @staticmethod
    def _check_headers(headers: Mapping[str, str]) -> dict[str, str]:
        for name, value in headers.items():
            if _UNSAFE_HEADER.search(str(name)) or _UNSAFE_HEADER.search(str(value)):
                raise InvalidRequestError(f"Cabeçalho {name!r} com quebra de linha.")
        return {str(k): str(v) for k, v in headers.items()}

    def url(self, path: str, path_params: Mapping[str, Any], query: Iterable[tuple[str, Any, bool]]) -> str:
        for name, value in path_params.items():
            if not isinstance(value, str) or value == "":
                raise InvalidRequestError(f"Parâmetro {name!r} obrigatório (texto não vazio).")
            path = path.replace("{" + name + "}", urllib.parse.quote(value, safe=""))
        pairs: list[tuple[str, str]] = []
        for wire, value, repeat in query:
            if value is None:
                continue
            if repeat:
                values = [value] if isinstance(value, str) else list(value)
                pairs.extend((wire, _scalar(v)) for v in values)
            else:
                pairs.append((wire, _scalar(value)))
        suffix = ("?" + urllib.parse.urlencode(pairs)) if pairs else ""
        return self.base_url + path + suffix

    def request(
        self,
        method: str,
        path: str,
        *,
        path_params: Mapping[str, Any] | None = None,
        query: Iterable[tuple[str, Any, bool]] = (),
        json_body: Any = None,
        has_body: bool = False,
        files: Mapping[str, FileUpload] | None = None,
        idempotency: str | None = None,
        idempotency_key: str | None = None,
        timeout: float | None = None,
        headers: Mapping[str, str] | None = None,
        accept: str = "application/json",
    ) -> RawResponse:
        url = self.url(path, path_params or {}, query)
        final = dict(self.default_headers)
        final.update(self._check_headers(headers or {}))
        final["Authorization"] = f"Bearer {self.__token}"
        final["Accept"] = accept
        final["User-Agent"] = self.user_agent

        if idempotency_key is None and idempotency == "required":
            idempotency_key = str(uuid.uuid4())
        if idempotency_key is not None:
            if not IDEMPOTENCY_KEY.fullmatch(idempotency_key):
                raise InvalidRequestError("Idempotency-Key: de 1 a 255 caracteres ASCII visíveis, sem espaços.")
            final["Idempotency-Key"] = idempotency_key

        data: bytes | None = None
        if files is not None:
            data, content_type = self._multipart(files)
            final["Content-Type"] = content_type
        elif has_body and json_body is not None:
            data = json.dumps(json_body, ensure_ascii=False, separators=(",", ":")).encode("utf-8")
            final["Content-Type"] = "application/json"

        request = urllib.request.Request(url, data=data, method=method)
        for name, value in final.items():
            request.add_header(name, value)

        limit = self.timeout if timeout is None else float(timeout)
        try:
            with self._opener.open(request, timeout=limit) as response:
                status = int(response.status)
                response_headers = {k.lower(): v for k, v in response.headers.items()}
                body = response.read()
        except urllib.error.HTTPError as exc:
            status = int(exc.code)
            response_headers = {k.lower(): v for k, v in (exc.headers or {}).items()}
            try:
                body = exc.read()
            except OSError:
                body = b""
        except (socket.timeout, TimeoutError):
            raise RequestTimeoutError(f"Tempo esgotado ({limit:g} s) em {method} {path}.") from None
        except urllib.error.URLError as exc:
            if isinstance(exc.reason, (socket.timeout, TimeoutError)):
                raise RequestTimeoutError(f"Tempo esgotado ({limit:g} s) em {method} {path}.") from None
            raise NetworkError(f"Falha de conexão em {method} {path}: {exc.reason}") from None
        except OSError as exc:
            raise NetworkError(f"Falha de conexão em {method} {path}: {exc}") from None

        if status < 200 or status >= 300:
            raise ApiError.from_response(status, response_headers, body)
        return RawResponse(status, response_headers, body)

    @staticmethod
    def _multipart(files: Mapping[str, FileUpload]) -> tuple[bytes, str]:
        boundary = "AssinaVeloxSdk" + uuid.uuid4().hex
        chunks: list[bytes] = []
        for name, upload in files.items():
            if _UNSAFE_HEADER.search(upload.content_type):
                raise InvalidRequestError("content_type do arquivo com quebra de linha.")
            chunks.append(f"--{boundary}\r\n".encode("ascii"))
            chunks.append(
                (
                    f'Content-Disposition: form-data; name="{_quote_param(name)}"; '
                    f'filename="{_quote_param(upload.filename)}"\r\n'
                    f"Content-Type: {upload.content_type}\r\n\r\n"
                ).encode("utf-8")
            )
            chunks.append(bytes(upload.content))
            chunks.append(b"\r\n")
        chunks.append(f"--{boundary}--\r\n".encode("ascii"))
        return b"".join(chunks), f"multipart/form-data; boundary={boundary}"


def data_of(response: RawResponse) -> Any:
    body = response.json()
    return body.get("data") if isinstance(body, dict) else None


def meta_of(response: RawResponse) -> dict[str, Any]:
    body = response.json()
    meta = body.get("meta") if isinstance(body, dict) else None
    return meta if isinstance(meta, dict) else {}

File: sdks/node/src/http.ts
import { randomUUID } from 'node:crypto';

import {
    ApiError,
    InvalidRequestError,
    NetworkError,
    RequestTimeoutError,
} from './errors.js';
import { toBytes, type FileUpload } from './files.js';
import { API_MAJOR, SDK_VERSION } from './version.js';

/** Opções do cliente. */
export interface ClientOptions {
    /** Endereço da sua instalação, terminando em `/api/v1`. */
    baseUrl: string;
    /** Texto da chave criada em Integrações → Chaves (exibido uma única vez). */
    token: string;
    /** Tempo máximo de cada requisição, em milissegundos (padrão: 30000). */
    timeoutMs?: number;
    /** Cabeçalhos extras em todas as requisições. */
    headers?: Record<string, string>;
    /** `fetch` alternativo (testes, proxies). Padrão: o `fetch` global do Node 18+. */
    fetch?: typeof fetch;
}

/** Opções de uma chamada. */
export interface RequestOptions {
    /**
     * 1 a 255 caracteres ASCII visíveis. Nas criações e no envio, sem chave o
     * SDK gera um UUID v4.
     */
    idempotencyKey?: string;
    /** Milissegundos; ausente usa o do cliente. */
    timeoutMs?: number;
    /** Cabeçalhos extras (não substituem Authorization, Accept nem User-Agent). */
    headers?: Record<string, string>;
    /** Cancela a requisição. */
    signal?: AbortSignal;
}

export interface RawResponse {
    status: number;
    headers: Record<string, string>;
    body: Uint8Array;
}

export interface RequestSpec {
    method: string;
    path: string;
    pathParams?: Record<string, string>;
    query?: object;
    querySpec?: Record<string, readonly [string, boolean]>;
    json?: unknown;
    hasBody?: boolean;
    file?: FileUpload;
    fileField?: string;
    idempotency?: 'required' | 'optional';
    options?: RequestOptions;
    accept?: string;
}

const IDEMPOTENCY_KEY = /^[\x21-\x7e]{1,255}$/;
const UNSAFE = /[\r\n\0]/;
const RESERVED = new Set([
    'authorization',
    'accept',
    'user-agent',
    'idempotency-key',
    'content-type',
    'content-length',
    'host',
]);

function checkHeaders(
    headers: Record<string, string> | undefined,
): Record<string, string> {
    const result: Record<string, string> = {};
    for (const [name, value] of Object.entries(headers ?? {})) {
        if (UNSAFE.test(name) || UNSAFE.test(String(value))) {
            throw new InvalidRequestError(
                `Cabeçalho "${name}" com quebra de linha.`,
            );
        }
        result[name] = String(value);
    }
    return result;
}

function scalar(name: string, value: unknown): string {
    if (typeof value === 'boolean') {
        return value ? 'true' : 'false';
    }
    if (typeof value === 'number' || typeof value === 'string') {
        return String(value);
    }
    throw new InvalidRequestError(
        `Parâmetro de consulta "${name}" com tipo não suportado.`,
    );
}

function concat(parts: Uint8Array[]): Uint8Array {
    const total = parts.reduce((sum, part) => sum + part.byteLength, 0);
    const result = new Uint8Array(total);
    let offset = 0;
    for (const part of parts) {
        result.set(part, offset);
        offset += part.byteLength;
    }
    return result;
}

function multipart(
    field: string,
    file: FileUpload,
): { body: Uint8Array; contentType: string } {
    const contentType = file.contentType ?? 'application/octet-stream';
    if (UNSAFE.test(contentType)) {
        throw new InvalidRequestError(
            'contentType do arquivo com quebra de linha.',
        );
    }
    const boundary = `AssinaVeloxSdk${randomUUID().replaceAll('-', '')}`;
    const quote = (value: string): string =>
        value
            .replaceAll('"', '%22')
            .replaceAll('\r', '%0D')
            .replaceAll('\n', '%0A');
    const encoder = new TextEncoder();
    const head =
        `--${boundary}\r\n` +
        `Content-Disposition: form-data; name="${quote(field)}"; filename="${quote(file.filename)}"\r\n` +
        `Content-Type: ${contentType}\r\n\r\n`;
    return {
        body: concat([
            encoder.encode(head),
            toBytes(file.content),
            encoder.encode(`\r\n--${boundary}--\r\n`),
        ]),
        contentType: `multipart/form-data; boundary=${boundary}`,
    };
}

/**
 * Núcleo HTTP do SDK (uso interno). O token fica num campo privado (#): não
 * aparece em `console.log`, `util.inspect` nem em mensagens de erro.
 * Redirecionamentos não são seguidos.
 */
export class HttpClient {
    readonly baseUrl: string;
    readonly timeoutMs: number;
    readonly userAgent: string;
    readonly #token: string;
    readonly #headers: Record<string, string>;
    readonly #fetch: typeof fetch;

    constructor(options: ClientOptions) {
        if (
            typeof options.baseUrl !== 'string' ||
            !/^https?:\/\/[^/\s]+/i.test(options.baseUrl)
        ) {
            throw new InvalidRequestError(
                'baseUrl precisa ser http(s)://…/api/v1 da sua instalação.',
            );
        }
        // Só caracteres visíveis de um byte (a regra de valor de cabeçalho HTTP): NUL, controle,
        // espaço ou acima de 0xFF fariam o `fetch` recusar o cabeçalho com uma mensagem que
        // repete o valor — e o token iria para a exceção.
        if (
            typeof options.token !== 'string' ||
            options.token === '' ||
            /[^\x21-\x7e\x80-\xff]/.test(options.token)
        ) {
            throw new InvalidRequestError(
                'token vazio ou com espaço: use o texto exibido na criação da chave.',
            );
        }
        const timeoutMs = options.timeoutMs ?? 30_000;
        if (!(timeoutMs > 0)) {
            throw new InvalidRequestError(
                'timeoutMs precisa ser maior que zero.',
            );
        }
        const fetchImpl = options.fetch ?? globalThis.fetch;
        if (typeof fetchImpl !== 'function') {
            throw new InvalidRequestError(
                'fetch indisponível: use Node.js 18 ou mais novo.',
            );
        }
        this.baseUrl = options.baseUrl.replace(/\/+$/, '');
        this.timeoutMs = timeoutMs;
        this.#token = options.token;
        this.#headers = checkHeaders(options.headers);
        this.#fetch = fetchImpl;
        this.userAgent = `assinavelox-node/${SDK_VERSION} (api-v${API_MAJOR}; node/${process.version})`;
    }

    url(
        path: string,
        pathParams: Record<string, string>,
        query: object,
        querySpec: Record<string, readonly [string, boolean]>,
    ): string {
        let resolved = path;
        for (const [name, value] of Object.entries(pathParams)) {
            if (typeof value !== 'string' || value === '') {
                throw new InvalidRequestError(
                    `Parâmetro "${name}" obrigatório (texto não vazio).`,
                );
            }
            resolved = resolved.replace(`{${name}}`, encodeURIComponent(value));
        }
        const values = query as Record<string, unknown>;
        const unknown = Object.keys(values).filter(
            (name) => !(name in querySpec),
        );
        if (unknown.length > 0) {
            throw new InvalidRequestError(
                `Parâmetro de consulta desconhecido: ${unknown.join(', ')}.`,
            );
        }
        const pairs: string[] = [];
        for (const [name, [wire, repeat]] of Object.entries(querySpec)) {
            const value = values[name];
            if (value === undefined || value === null) {
                continue;
            }
            const items = repeat && Array.isArray(value) ? value : [value];
            for (const item of items) {
                pairs.push(
                    `${encodeURIComponent(wire)}=${encodeURIComponent(scalar(name, item))}`,
                );
            }
        }
        return (
            this.baseUrl +
            resolved +
            (pairs.length > 0 ? `?${pairs.join('&')}` : '')
        );
    }

    async request(spec: RequestSpec): Promise<RawResponse> {
        const options = spec.options ?? {};
        const url = this.url(
            spec.path,
            spec.pathParams ?? {},
            spec.query ?? {},
            spec.querySpec ?? {},
        );

        const headers: Record<string, string> = {};
        for (const [name, value] of Object.entries({
            ...this.#headers,
            ...checkHeaders(options.headers),
        })) {
            if (!RESERVED.has(name.toLowerCase())) {
                headers[name.toLowerCase()] = value;
            }
        }
        headers.authorization = `Bearer ${this.#token}`;
        headers.accept = spec.accept ?? 'application/json';
        headers['user-agent'] = this.userAgent;

        let key = options.idempotencyKey;
        if (key === undefined && spec.idempotency === 'required') {
            key = randomUUID();
        }
        if (key !== undefined) {
            if (!IDEMPOTENCY_KEY.test(key)) {
                throw new InvalidRequestError(
                    'Idempotency-Key: de 1 a 255 caracteres ASCII visíveis, sem espaços.',
                );
            }
            headers['idempotency-key'] = key;
        }

        let body: Uint8Array | undefined;
        if (spec.file !== undefined) {
            const encoded = multipart(spec.fileField ?? 'file', spec.file);
            body = encoded.body;
            headers['content-type'] = encoded.contentType;
        } else if (
            spec.hasBody === true &&
            spec.json !== undefined &&
            spec.json !== null
        ) {
            body = new TextEncoder().encode(JSON.stringify(spec.json));
            headers['content-type'] = 'application/json';
        }

        const limit = options.timeoutMs ?? this.timeoutMs;
        const controller = new AbortController();
        let timedOut = false;
        const timer = setTimeout(() => {
            timedOut = true;
            controller.abort();
        }, limit);
        const external = options.signal;
        const relay = (): void => controller.abort(external?.reason);
        if (external?.aborted === true) {
            controller.abort(external.reason);
        } else {
            external?.addEventListener('abort', relay, { once: true });
        }

        const where = `${spec.method} ${spec.path}`;
        try {
            const response = await this.#fetch(url, {
                method: spec.method,
                headers,
                body,
                redirect: 'manual',
                signal: controller.signal,
            });
            const payload = new Uint8Array(await response.arrayBuffer());
            const responseHeaders: Record<string, string> = {};
            response.headers.forEach((value, name) => {
                responseHeaders[name.toLowerCase()] = value;
            });
            if (response.status < 200 || response.status >= 300) {
                throw ApiError.fromResponse(
                    response.status,
                    responseHeaders,
                    payload,
                );
            }
            return {
                status: response.status,
                headers: responseHeaders,
                body: payload,
            };
        } catch (error) {
            if (error instanceof ApiError) {
                throw error;
            }
            if (timedOut) {
                throw new RequestTimeoutError(
                    `Tempo esgotado (${limit} ms) em ${where}.`,
                );
            }
            if (external?.aborted === true) {
                throw error;
            }
            const cause = (
                error as { cause?: { code?: string; message?: string } }
            ).cause;
            let reason = String(
                cause?.code ?? cause?.message ?? (error as Error).message,
            );
            // Nunca repetir um cabeçalho recusado (o valor seria "Bearer <token>") nem o token:
            // nesses casos a mensagem é genérica e o erro original NÃO vai como `cause`.
            const unsafe =
                (error instanceof TypeError && /header/i.test(reason)) ||
                reason.includes(this.#token);
            if (unsafe) {
                reason = 'cabeçalho inválido';
            }
            throw new NetworkError(
                `Falha de conexão em ${where}: ${reason}`,
                unsafe ? undefined : { cause: error },
            );
        } finally {
            clearTimeout(timer);
            external?.removeEventListener('abort', relay);
        }
    }
}

export function jsonOf(response: RawResponse): unknown {
    if (response.body.byteLength === 0) {
        return null;
    }
    return JSON.parse(new TextDecoder().decode(response.body)) as unknown;
}

export function dataOf(response: RawResponse): unknown {
    const json = jsonOf(response);
    return json !== null && typeof json === 'object'
        ? (json as Record<string, unknown>).data
        : undefined;
}

export function metaOf(response: RawResponse): Record<string, unknown> {
    const json = jsonOf(response);
    const meta =
        json !== null && typeof json === 'object'
            ? (json as Record<string, unknown>).meta
            : undefined;
    return meta !== null && typeof meta === 'object' && !Array.isArray(meta)
        ? (meta as Record<string, unknown>)
        : {};
}
